src/lib/group.ts

```diff
@@ -0,0 +1,364 @@
+    x: FieldVar | Field | number | string | bigint;
+    y: FieldVar | Field | number | string | bigint;
+  }) {
+    this.x = isField(x) ? x : new Field(x);
+    this.y = isField(y) ? y : new Field(y);
+
+    if (this.#isConstant()) {
+      const { add, mul, square } = Fp;
+
+      let x_bigint = this.x.toBigInt();
+      let y_bigint = this.y.toBigInt();
+
+      let onCurve =
+        add(mul(x_bigint, mul(x_bigint, x_bigint)), Pallas.b) ===
+        square(y_bigint);
+
+      if (!onCurve) {
+        throw Error(
+          `(x: ${x_bigint}, y: ${y_bigint}) is not a valid group element`
+        );
+      }
+    }
+  }
+
+  // helpers
+  static #fromAffine({ x, y }: { x: bigint; y: bigint; infinity: boolean }) {
+    return new Group({ x, y });
+  }
+
+  static #fromProjective({ x, y, z }: { x: bigint; y: bigint; z: bigint }) {
+    return this.#fromAffine(Pallas.toAffine({ x, y, z }));
+  }
+
+  #toTuple(): [0, FieldVar, FieldVar] {
+    return [0, this.x.value, this.y.value];
+  }
+
+  #isConstant() {
+    return this.x.isConstant() && this.y.isConstant();
+  }
+
+  #toAffine() {
+    return Pallas.fromAffine({
+      x: this.x.toBigInt(),
+      y: this.y.toBigInt(),
+      infinity: false,
+    });
+  }
+
+  /**
+   * Adds this {@link Group} element to another {@link Group} element.
+   *
+   * ```ts
+   * let g1 = Group({ x: 1, y: 1})
+   * let g2 = g1.add(g1)
+   * ```
+   */
+  add(g: Group) {
+    if (this.#isConstant() && g.#isConstant()) {
+      if (this.x.toBigInt() === 0n) {
+        return g;
+      } else if (g.x.toBigInt() === 0n) {
+        return this;
+      } else {
+        let g_proj = Pallas.add(this.#toAffine(), g.#toAffine());
+        return Group.#fromProjective(g_proj);
+      }
+    } else {
+      let [, x, y] = Snarky.group.add(this.#toTuple(), g.#toTuple());
+      return new Group({ x, y });
+    }
+  }
+
+  /**
+   * Subtracts another {@link Group} element from this one.
+   */
+  sub(y: Group) {
+    return this.add(y.neg());
+  }
+
+  /**
+   * Negates this {@link Group}. Under the hood, it simply negates the `y` coordinate and leaves the `x` coordinate as is.
+   */
+  neg() {
+    let { x, y } = this;
+    return new Group({ x, y: y.neg() });
+  }
+
+  /**
+   * Elliptic curve scalar multiplication. Scales the {@link Group} element `n`-times by itself, where `n` is the {@link Scalar}.
+   *
+   * ```typescript
+   * let s = Scalar(5);
+   * let 5g = g.scale(s);
+   * ```
+   */
+  scale(s: Scalar | number | bigint) {
+    let scalar =
+      typeof s === 'bigint' || typeof s === 'number'
+        ? Scalar.fromBigInt(BigInt(s))
+        : s;
+    let fields = scalar.toFields();
+
+    if (this.#isConstant() && fields.every((f) => f.isConstant())) {
+      let g_proj = Pallas.scale(this.#toAffine(), BigInt(scalar.toJSON()));
+      return Group.#fromProjective(g_proj);
+    } else {
+      let [, x, y] = Snarky.group.scale(this.#toTuple(), [
+        0,
+        ...fields.map((f) => f.value).reverse(),
```

Comment:
while I appreciate keeping the `Snarky` interface low-level and having logic like reversing the scalar bits on the JS side, it would be good if a doccomment on `Snarky.group.scale` documented this unexpected behaviour